src/blocked.c

```diff
@@ -174,57 +198,157 @@ void queueClientForReprocessing(client *c) {
         listDelNode(server.postponed_clients,c->postponed_list_node);
         c->postponed_list_node = NULL;
-    } else if (c->btype == BLOCKED_SHUTDOWN) {
+    } else if (c->bstate.btype == BLOCKED_SHUTDOWN) {
         /* No special cleanup. */
     } else {
         serverPanic("Unknown btype in unblockClient().");
     }
 
-    /* Reset the client for a new query since, for blocking commands
-     * we do not do it immediately after the command returns (when the
-     * client got blocked) in order to be still able to access the argument
-     * vector from module callbacks and updateStatsOnUnblock. */
-    if (c->btype != BLOCKED_POSTPONE && c->btype != BLOCKED_SHUTDOWN) {
+    /* Reset the client for a new query, unless the client has pending command to process
+     * or in case a shutdown operation was canceled and we are still in the processCommand sequence  */
+    if (!(c->flags & CLIENT_PENDING_COMMAND) && c->bstate.btype != BLOCKED_SHUTDOWN) {
         freeClientOriginalArgv(c);
         resetClient(c);
     }
 
     /* Clear the flags, and put the client in the unblocked list so that
      * we'll process new commands in its query buffer ASAP. */
     server.blocked_clients--;
-    server.blocked_clients_by_type[c->btype]--;
+    server.blocked_clients_by_type[c->bstate.btype]--;
     c->flags &= ~CLIENT_BLOCKED;
-    c->btype = BLOCKED_NONE;
+    c->bstate.btype = BLOCKED_NONE;
     removeClientFromTimeoutTable(c);
     queueClientForReprocessing(c);
 }
 
+/* Unblock a client once a specific key became available for it.
+ * This function will remove the client from the list of clients blocked on this key
+ * and also remove the key from the dictionary of keys this client is blocked on.
+ * in case the client has a command pending it will process it immediately.  */
+void unblockClientOnKey(client *c, robj *key) {
+    list *l;
+    bkinfo *bki;
+
+    bki = dictFetchValue(c->bstate.keys, key);
+    /* Remove this client from the list of clients waiting for this key. */
+    l = dictFetchValue(c->db->blocking_keys,key);
+    serverAssertWithInfo(c,key,l != NULL);
+    listDelNode(l,bki->listnode);
+    /* If the list is empty we need to remove it to avoid wasting memory */
+    if (listLength(l) == 0)
+        dictDelete(c->db->blocking_keys,key);
+    dictDelete(c->bstate.keys,key);
+
+
+    /* Client is no more blocked on keys */
+    if (dictSize(c->bstate.keys) == 0 ||
+        // in case of blocking API calls, we might be blocked on several keys.
+        // however we should force unblock the entire blocking keys
+        c->bstate.btype == BLOCKED_STREAM ||
+        c->bstate.btype == BLOCKED_LIST ||
+        c->bstate.btype == BLOCKED_ZSET) {
+
+        unblockClient(c);
+        /* In case this client was blocked on keys during command
+         * we need to
+         */
+        if (c->flags & CLIENT_PENDING_COMMAND) {
+            c->flags &= ~CLIENT_PENDING_COMMAND;
+            processCommandAndResetClient(c);
```

Comment:
I don't think we want to recursively call this API. For LMOVE, you could have a chain of calls that goes deeper into the stack. Instead, we can probably just loop until all of the unblocked keys have been satisfied. Which I think is handled properly in handleClientsBlockedOnKey, we just need to make sure we don't enter that function recursively.

Reply:
I am not sure I understand. removing the call here is basically a breaking change. we currently depend on the logic which execute a blocked command as SOON as the key is ready...

Reply:
I'm not sure this actually changes the behavior. Instead of recursing to do the unblocking we would loop over them in a while loop. It's like a tail recursion optimization. I think we generally want to avoid recursion inside Redis commands.

Reply:
The way I see it is that the current behavior require us to make sure no other client will be able to successfully run a blocking command before an already blocked client has processed it, so this recursion (which is not new) was left as it is.
I also thought of this while writing the refactor, and IMO the solution is that a client will also check on each key of a blocking command, if that key has pending blocked clients (in which case it should also be blocked). 
I agree that this recursion is not preferred, but I would like to leave some optimizations to the next phase, as this is becoming  hard to approve since the code change is large.